chronos/timer.nim

```diff
@@ -8,45 +8,397 @@
+
+  Duration* = object
+    value: int64
+
+  DurationOrMoment* = Duration | Moment
+
+func `+`*(a: Duration, b: Duration): Duration {.inline.} =
+  ## Duration + Duration = Duration
+  result.value = a.value + b.value
+
+func `+`*(a: Duration, b: Moment): Moment {.inline.} =
+  ## Duration + Moment = Moment
+  result.value = a.value + b.value
+
+func `+`*(a: Moment, b: Duration): Moment {.inline.} =
+  ## Moment + Duration = Moment
+  result.value = a.value + b.value
+
+func `+=`*(a: var Moment, b: Duration) {.inline.} =
+  ## Moment += Duration
+  a.value += b.value
+
+func `+=`*(a: var Duration, b: Duration) {.inline.} =
+  ## Duration += Duration
+  a.value += b.value
+
+func `-`*(a, b: Moment): Duration {.inline.} =
+  ## Moment - Moment = Duration
+  ##
+  ## Note: Duration can't be negative.
+  result.value = if a.value >= b.value: a.value - b.value else: 0'i64
+
+func `-`*(a: Moment, b: Duration): Moment {.inline.} =
+  ## Moment - Duration = Moment
+  ##
+  ## Note: Moment can be negative
+  result.value = a.value - b.value
+
+func `-`*(a: Duration, b: Duration): Duration {.inline.} =
+  ## Duration - Duration = Duration
+  ##
+  ## Note: Duration can't be negative.
+  result.value = if a.value >= b.value: a.value - b.value else: 0'i64
+
+func `-=`*(a: var Duration, b: Duration): Duration {.inline.} =
+  ## Duration -= Duration
+  a.value = if a.value >= b.value: a.value - b.value else: 0'i64
+
+func `-=`*(a: var Moment, b: Duration): Moment {.inline.} =
+  ## Moment -= Duration
+  a.value -= b.value
+
+func `==`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` equal to ``b``.
+  result = (a.value == b.value)
+
+func `==`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` equal to ``b``.
+  result = (a.value == b.value)
+
+func `<`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less then ``b``.
+  result = (a.value < b.value)
+
+func `<`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less then ``b``.
+  result = (a.value < b.value)
+
+func `<=`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less or equal ``b``.
+  result = (a.value <= b.value)
+
+func `<=`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less or equal ``b``.
+  result = (a.value <= b.value)
+
+func `>`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger then ``b``.
+  result = (a.value > b.value)
+
+func `>`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger then ``b``.
+  result = (a.value > b.value)
+
+func `>=`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger or equal ``b``.
+  result = (a.value >= b.value)
+
+func `>=`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger or equal ``b``.
+  result = (a.value >= b.value)
+
+func `*`*(a, b: Duration): Duration {.inline.} =
+  result.value = a.value * b.value
+
+func `div`*(a, b: Duration): Duration {.inline.} =
```

Comment:
this should probably return a scalar value (int64) that is the ratio between the two durations.. a division would look like: `div(a: Duration, b: SomeInteger): int64`